nilearn/maskers/nifti_masker.py

```diff
@@ -615,7 +616,16 @@ def transform_single_imgs(
             confounds=confounds,
             sample_mask=sample_mask,
             copy=copy,
-            dtype=self.dtype,
         )
 
-        return data
+        if not isinstance(data, np.ndarray):
+            # in case data is a cached MemorizedResult
+            # only happens when _shelving is True
+            return data
+
+        imgs = load_img(imgs)
+        target_dtype = get_target_dtype(img_data_dtype(imgs), self.dtype)
+        if target_dtype is None:
+            target_dtype = img_data_dtype(imgs)
```

Comment:
for this masker and several others the new type is set after extracting / filtering the data rather than during this process

but only for this Masker and MultiMasker do we need extra check to deal with potentially MemorizedResult because only for those we use and test behavior with _shelving

https://github.com/nilearn/nilearn/issues/5523

Reply:
You mean, just add additional checks that would detect a potential caching issue that would result in wrong dtype ? If it's bot too much of a hassle, this makes sense.

Reply:
LMK when you're done.

Reply:
I have added some extra tests to make sure things are the same when caching